### PR TITLE
server pod fail- multipart tests

### PR DESCRIPTION
**Pull request by sonalk0209**
# Problem Statement
Multipart tests for server pod failure

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/ha/test_server_pod_failure.py [hunk 1]
```diff
@@ -1590,3 +1590,571 @@ def test_copy_obj_during_server_pod_shutdown(self):
...
+                                  f"Failed to match checksum: {upload_checksum},"
+                                  f" {download_checksum}")
+        LOGGER.info("Matched checksum: %s, %s", upload_checksum, download_checksum)
+        LOGGER.info("Step 6: Successfully downloaded the object and verified the checksum")
+
+        LOGGER.info("Removing files %s and %s", self.multipart_obj_path, download_path)
+        sysutils.remove_file(self.multipart_obj_path)
+        sysutils.remove_file(download_path)
+
+        LOGGER.info("Step 7: Create new bucket and multipart upload and then download 5GB object")
```
Comment (ankita-93) [inline tests/ha/test_server_pod_failure.py:1814]: same as above

File: tests/ha/test_server_pod_failure.py [hunk 2]
```diff
@@ -1590,3 +1590,571 @@ def test_copy_obj_during_server_pod_shutdown(self):
...
+        pod_list.remove(pod_name)
+        LOGGER.info("Step 5: Check services status on remaining pods %s are in online state",
+                    pod_list)
+        resp = self.hlth_master_list[0].get_pod_svc_status(pod_list=pod_list, fail=False)
+        LOGGER.debug("Response: %s", resp)
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 5: Services status on remaining pods %s are in online state", pod_list)
+
+        LOGGER.info("Step 6: Checking response from background process")
+        thread.join()
```
Comment (ankita-93) [inline tests/ha/test_server_pod_failure.py:1929]: We need to clear an event too.

File: tests/ha/test_server_pod_failure.py [hunk 3]
```diff
@@ -1590,3 +1590,571 @@ def test_copy_obj_during_server_pod_shutdown(self):
...
+        etag_list = parts_etag1 + parts_etag2
+        parts_etag = sorted(etag_list, key=lambda d: d['PartNumber'])
+
+        LOGGER.info("Step 8: Listing parts of multipart upload")
+        res = s3_mp_test_obj.list_parts(mpu_id, self.bucket_name, self.object_name)
+        assert_utils.assert_true(res[0], res)
+        assert_utils.assert_equal(len(res[1]["Parts"]), total_parts)
+        LOGGER.info("Step 8: Listed parts of multipart upload. Count: %s", len(res[1]["Parts"]))
+
+        LOGGER.info("Step 9: Completing multipart upload Completing multipart upload and check "
```
Comment (ankita-93) [inline tests/ha/test_server_pod_failure.py:2132]: Completing multipart upload is twice here..

File: tests/ha/test_server_pod_failure.py [hunk 4]
```diff
@@ -1590,3 +1590,574 @@ def test_copy_obj_during_server_pod_shutdown(self):
...
+
+        LOGGER.info("Step 2: Shutdown the server pod safely by making replicas=0")
+        LOGGER.info("Get pod name to be deleted")
+        pod_list = self.node_master_list[0].get_all_pods(pod_prefix=const.SERVER_POD_NAME_PREFIX)
+        pod_name = random.sample(pod_list, 1)[0]
+        hostname = self.node_master_list[0].get_pod_hostname(pod_name=pod_name)
+        LOGGER.info("Deleting pod %s", pod_name)
+        resp = self.node_master_list[0].create_pod_replicas(num_replica=0, pod_name=pod_name)
+        assert_utils.assert_false(resp[0], f"Failed to delete pod {pod_name} by making replicas=0")
+        LOGGER.info("Step 2: Successfully shutdown/deleted pod %s by making replicas=0", pod_name)
```
Comment (rahul-hatwar) [inline tests/ha/test_server_pod_failure.py:1647]: shutdown/delete server pod

**Review by ApurwaMohite — APPROVED** [review]

File: tests/ha/test_server_pod_failure.py [hunk 5]
```diff
@@ -1592,6 +1592,256 @@ def test_copy_obj_during_server_pod_shutdown(self):
...
+        LOGGER.info("Get pod name to be deleted")
+        pod_list = self.node_master_list[0].get_all_pods(pod_prefix=const.SERVER_POD_NAME_PREFIX)
+        pod_name = random.sample(pod_list, 1)[0]
+        hostname = self.node_master_list[0].get_pod_hostname(pod_name=pod_name)
+        LOGGER.info("Deleting pod %s", pod_name)
+        resp = self.node_master_list[0].delete_deployment(pod_name=pod_name)
+        LOGGER.debug("Response: %s", resp)
+        assert_utils.assert_false(resp[0], f"Failed to delete pod {pod_name} by deleting deployment"
+                                           " (unsafe)")
+        LOGGER.info("Step 2: Successfully shutdown/deleted pod %s by deleting deployment (unsafe)",
```
Comment (rahul-hatwar) [inline tests/ha/test_server_pod_failure.py:1773]: deleted server pod

File: tests/ha/test_server_pod_failure.py [hunk 6]
```diff
@@ -1662,6 +1912,171 @@ def test_ios_unsafe_server_pod_shutdown(self):
...
+        LOGGER.info("Step 5: Check services status on remaining pods %s are in online state",
+                    pod_list)
+        resp = self.hlth_master_list[0].get_pod_svc_status(pod_list=pod_list, fail=False)
+        LOGGER.debug("Response: %s", resp)
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 5: Services status on remaining pods %s are in online state", pod_list)
+
+        LOGGER.info("Step 6: Checking response from background process")
+        thread.join()
+        event.clear()
```
Comment (ankita-93) [inline tests/ha/test_server_pod_failure.py:2002]: This should go after line 1998. We have observed that thread join takes time to complete the whole task. We don't want event to be set till that point.

File: tests/ha/test_server_pod_failure.py [hunk 7]
```diff
@@ -1662,6 +1912,171 @@ def test_ios_unsafe_server_pod_shutdown(self):
...
+        pod_list = self.node_master_list[0].get_all_pods(pod_prefix=const.SERVER_POD_NAME_PREFIX)
+        pod_name = random.sample(pod_list, 1)[0]
+        hostname = self.node_master_list[0].get_pod_hostname(pod_name=pod_name)
+        LOGGER.info("Deleting pod %s", pod_name)
+        event.set()
+        resp = self.node_master_list[0].delete_deployment(pod_name=pod_name)
+        LOGGER.debug("Response: %s", resp)
+        assert_utils.assert_false(resp[0], f"Failed to delete pod {pod_name} by deleting deployment"
+                                           " (unsafe)")
+        LOGGER.info("Step 2: Successfully shutdown/deleted pod %s by deleting deployment (unsafe)",
```
Comment (rahul-hatwar) [inline tests/ha/test_server_pod_failure.py:1973]: server pod

**Review by ankita-93 — APPROVED** [review]
LGTM

**Review by rahul-hatwar — APPROVED** [review]
Minor comment.
LGTM

**Comment by sonalk0209** [comment]
[collect_pr1448.log](https://github.com/Seagate/cortx-test/files/8509881/collect_pr1448.log)
collections log

**Review by ravindrachoudhari — APPROVED** [review]

**Pull request merged**